### PR TITLE
Create EnpointSlices

### PR DESCRIPTION
**Pull request by aswinsuryan**
* An endpointslice will be created when a ServiceImport is created
for a headless service

* When a ServiceImport for headless service is created and endpoint
controller will be created which will listen for the endpoints with
same label that the service selector uses.

* The endpoint slice will have the corresponding ServiceImport set as
owner.

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

### CODE REVIEW

File: pkg/agent/controller/agent.go [hunk 1]
```diff
@@ -166,8 +177,16 @@ func (a *Controller) serviceExportToRemoteServiceImport(obj runtime.Object, op s
...
 	}
 
+	var importType lighthousev2a1.ServiceImportType
+
+	if svc.Spec.ClusterIP == "None" {
+		importType = lighthousev2a1.Headless
+	} else {
+		importType = lighthousev2a1.SuperclusterIP
+	}
+
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go:187]: This doesn't seem related to this PR.

File: pkg/agent/controller/endpointcontroller.go [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+				return c.lighthouseClient.LighthouseV2alpha1().ServiceImports(c.lighthouseNamespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.lighthouseClient.LighthouseV2alpha1().ServiceImports(c.lighthouseNamespace).Watch(options)
+			},
+		},
+		&lighthousev2a1.ServiceImport{},
+		0,
+		cache.Indexers{},
+	)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			if exists {
+				c.serviceImportCreatedOrUpdated(obj, key)
+			} else {
+				klog.Errorf("Does not exist with key %q from the cache %v: %v", key, c.serviceInformer.GetIndexer().ListKeys(), err)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+	serviceImportCreated := obj.(*lighthousev2a1.ServiceImport)
+	annotations := serviceImportCreated.ObjectMeta.Annotations
+	serviceNameSpace := annotations[OriginNamespace]
+	serviceName := annotations[OriginName]
+
+	service, err := c.kubeClientSet.CoreV1().Services(serviceNameSpace).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, err)
+		return
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+
+func (e *EndpointController) endPointCreatedOrUpdated(obj interface{}, key string) {
+	endPoints := obj.(*corev1.Endpoints)
+	endpointSliceName := endPoints.Name + "-" + e.clusterID
+	newEndPointSlice := e.endpointSliceFromEndpoints(endPoints)
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentEndpointSice, err := e.kubeClientSet.DiscoveryV1beta1().EndpointSlices(endPoints.Namespace).
+			Get(endpointSliceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("Retrieving EndpointSlice %s from NameSpace %s failed due to %v", endpointSliceName, endPoints.Namespace, err)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+		labelSelector := labels.Set(map[string]string{"app": matchLabels["app"]}).AsSelector()
+		err := c.kubeClientSet.DiscoveryV1beta1().EndpointSlices(si.Namespace).
+			DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector.String()})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("Deleting EndpointSlice for Service Import %v from NameSpace %s failed due to %v", si, si.Namespace, err)
+			return err
+		}
+		c.endpointController[key].endPointqueue.ShutDown()
+		close(c.endpointController[key].stopCh)
+		delete(c.endpointController, key)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("Deleting EndpointSlice for endPoints %s from NameSpace %s failed due to %v", endPoints, endPoints.Namespace, err)
+			return err
+		}
+		return nil
+	})
+	if retryErr != nil {
+		klog.Errorf("Deleting EndpointSlice for endPoints %s from NameSpace %s failed after retry %v",
+			endPoints.Name, endPoints.Namespace, retryErr)
+	}
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/agent.go [hunk 8]
```diff
@@ -26,8 +26,8 @@ import (
 const (
 	submarinerIpamGlobalIp = "submariner.io/globalIp"
 	serviceUnavailable     = "ServiceUnavailable"
-	originName             = "origin-name"
-	originNamespace        = "origin-namespace"
+	OriginName             = "origin-name"
+	OriginNamespace        = "origin-namespace"
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: pkg/agent/controller/agent.go [hunk 9]
```diff
@@ -55,16 +55,21 @@ func New(spec *AgentSpecification, cfg *rest.Config) (*Controller, error) {
 		return nil, fmt.Errorf("error creating dynamic client: %v", err)
 	}
 
-	return NewWithDetail(spec, syncerConf, restMapper, localClient, kubeClientSet, lighthouseClient, nil,
+	siController, err := NewController(spec, kubeClientSet, lighthouseClient)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: pkg/agent/controller/agent.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+
+		si, ok = tombstone.Obj.(*lighthousev2a1.ServiceImport)
+
+		if !ok {
+			klog.Errorf("Failed to convert deleted tombstone object %v  to serviceimport", tombstone.Obj)
+			return
+		}
+	}
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+		if currentEndpointSice != nil && currentEndpointSice.ObjectMeta.Name != "" {
+			if equality.Semantic.DeepEqual(currentEndpointSice, newEndPointSlice) {
+				return nil
+			}
+			_, err = e.kubeClientSet.DiscoveryV1beta1().EndpointSlices(endPoints.Namespace).Update(newEndPointSlice)
+		} else {
+			_, err = e.kubeClientSet.DiscoveryV1beta1().EndpointSlices(endPoints.Namespace).Create(newEndPointSlice)
+		}
+		if err != nil {
+			klog.Errorf("Creating EndpointSlice %s from NameSpace %s failed due to : %v", endpointSliceName, endPoints.Namespace, err)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = labelSelector.String()
+				return c.kubeClientSet.CoreV1().Endpoints(metav1.NamespaceAll).Watch(options)
+			},
+		},
+		&corev1.Endpoints{},
+		0,
+		cache.Indexers{},
+	)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Doesn't it share an informer since we use SharedInformer? I just wanted to listen only for the endpoints which have ServiceExport created using the label filters.

File: pkg/agent/controller/endpointcontroller.go [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+}
+
+// Each EndpointController listens for the endpoints that backs a service and have a ServiceImport
+// It will create an endpoint slice corresponding to an endpoint object and set the owner references
+// to ServiceImport. The app label from the endpoint will be added to endpoint slice as well.
+type EndpointController struct {
+	kubeClientSet     kubernetes.Interface
+	endpointInformer  cache.SharedIndexInformer
+	endPointqueue     workqueue.RateLimitingInterface
+	serviceImportuid  types.UID
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+)
+
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		matchLabels := si.ObjectMeta.Labels
+		labelSelector := labels.Set(map[string]string{"app": matchLabels["app"]}).AsSelector()
+		err := c.kubeClientSet.DiscoveryV1beta1().EndpointSlices(si.Namespace).
+			DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector.String()})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("Deleting EndpointSlice for Service Import %v from NameSpace %s failed due to %v", si, si.Namespace, err)
+			return err
+		}
+		c.endpointController[key].endPointqueue.ShutDown()
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+
+	go c.serviceInformer.Run(stopCh)
+	go c.runServiceImportWorker()
+
+	go c.Stop(stopCh)
+
+	return nil
+}
+
+func (c *ServiceImportController) Stop(stopCh <-chan struct{}) {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/endpointcontroller.go [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+
+		endPoints, ok = tombstone.Obj.(*corev1.Endpoints)
+
+		if !ok {
+			klog.Errorf("Failed to convert deleted tombstone object %v  to endPoints", tombstone.Obj)
+			return
+		}
+	}
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpointcontroller.go]: Done.

File: pkg/agent/controller/serviceimport.go [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+			return err
+		}
+		isFound = true
+		return nil
+	})
+
+	if retryErr != nil || !isFound {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, retryErr)
+		return
+	}
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:172]: If the `Service` doesn't exist I guess we can ignore it since we should get a `ServiceImport` delete at some point. Otherwise we should log the error and re-queue.

File: pkg/agent/controller/serviceimport.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+		klog.Errorf("The service %q in the namespace %q without selector is not supported", serviceName, serviceNameSpace)
+		return
+	}
+
+	labelSelector := labels.Set(service.Spec.Selector).AsSelector()
+	endpointController, err := NewEndpointController(c.kubeClientSet, serviceImportCreated.ObjectMeta.UID,
+		serviceImportCreated.ObjectMeta.Name, c.clusterID)
+
+	if err != nil {
+		klog.Errorf("Creating endpoint controller for service %q in the namespace %q failed", serviceName, serviceNameSpace)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:184]:
```suggestion
		klog.Errorf("Error creating Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
```

File: pkg/agent/controller/serviceimport.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+		return
+	}
+
+	err = endpointController.Start(endpointController.stopCh, labelSelector)
+	if err != nil {
+		klog.Errorf("Starting endpoint controller for service %q in the namespace %q failed", serviceName, serviceNameSpace)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:190]:
```suggestion
		klog.Errorf(""Error starting Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
```

File: pkg/agent/controller/serviceimport.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+		return nil
+	})
+
+	if retryErr != nil || !isFound {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, retryErr)
+		return
+	}
+
+	if service.Spec.Selector == nil {
+		klog.Errorf("The service %q in the namespace %q without selector is not supported", serviceName, serviceNameSpace)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:175]:
```suggestion
		klog.Errorf("The service %s/%s without a Selector is not supported", serviceNameSpace, serviceName)
```

File: pkg/agent/controller/serviceimport.go [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+		return
+	}
+
+	c.endpointControllerCreated.Store(key, endpointController)
+}
+
+func (c *ServiceImportController) serviceImportDeleted(key string) {
+	obj, found := c.endpointControllerDeleted.Load(key)
+	if !found {
+		klog.Errorf("deleting endpointSlice for key %s failed since object not found", key)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:200]:
```suggestion
		klog.Warningf("No endpoint controller found  for %q", key)
```

File: pkg/agent/controller/serviceimport.go [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+			} else {
+				c.serviceImportDeleted(key)
+			}
+		}()
+	}
+}
+
+func (c *ServiceImportController) serviceImportCreatedOrUpdated(obj interface{}, key string) {
+	if _, found := c.endpointControllerCreated.Load(key); found {
+		klog.V(2).Infof("The endpointController is already running for  key %q", key)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:141]:
```suggestion
		klog.V(log.DEBUG).Infof("The endpoint controller is already running fo %q", key)
```r

File: pkg/agent/controller/types.go [hunk 24]
```diff
@@ -23,3 +28,36 @@ type AgentSpecification struct {
...
+
+// The ServiceImportController listens for a service import created in the submariner-operator namespace
+// and creates an EndpointController in response. The EndpointController will use the app label as filter
+// to listen only for the endpoints event related to ServiceImport created
+type ServiceImportController struct {
+	kubeClientSet             kubernetes.Interface
+	lighthouseClient          lighthouseClientset.Interface
+	serviceInformer           cache.SharedIndexInformer
+	queue                     workqueue.RateLimitingInterface
+	endpointControllerCreated sync.Map
```
Comment (tpantelis) [inline pkg/agent/controller/types.go:40]:
```suggestion
	endpointControllers sync.Map
```

File: pkg/agent/controller/serviceimport.go [hunk 25]
```diff
@@ -0,0 +1,224 @@
...
+		c.endpointControllerDeleted.Store(key, si)
+		c.queue.AddRateLimited(key)
+
+		return
+	}
+
+	if obj, found := c.endpointControllerCreated.Load(key); found {
+		endpointController := obj.(*EndpointController)
+		endpointController.endPointqueue.ShutDown()
+		close(endpointController.stopCh)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:221]: I suggest to move these into a `Stop` method on `EndpointController`.

File: pkg/agent/controller/types.go [hunk 26]
```diff
@@ -23,3 +28,36 @@ type AgentSpecification struct {
...
+	endpointControllerDeleted sync.Map
```
Comment (tpantelis) [inline pkg/agent/controller/types.go:41]:
```suggestion
	serviceImportDeleted sync.Map
```

File: pkg/agent/controller/serviceimport.go [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+
+					if !ok {
+						klog.Errorf("Failed to convert deleted tombstone object %v  to serviceimport", tombstone.Obj)
+						return
+					}
+				}
+				if si.Spec.Type != lighthousev2a1.Headless {
+					return
+				}
+				c.endpointControllerDeleted.Store(si, key)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:87]:
```suggestion
				c.endpointControllerDeleted.Store(key, si)
```

File: pkg/agent/controller/serviceimport.go [hunk 28]
```diff
@@ -0,0 +1,224 @@
...
+
+		return
+	}
+
+	if obj, found := c.endpointControllerCreated.Load(key); found {
+		endpointController := obj.(*EndpointController)
+		endpointController.endPointqueue.ShutDown()
+		close(endpointController.stopCh)
+		c.endpointControllerCreated.Delete(key)
+	}
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:223]: I think we should stop the controller regardless of whether or not the delete above fails, ie move this before the delete.

File: pkg/agent/controller/types.go [hunk 29]
```diff
@@ -23,3 +28,36 @@ type AgentSpecification struct {
 	Namespace        string
 	GlobalnetEnabled bool `split_words:"true"`
 }
+
+// The ServiceImportController listens for a service import created in the submariner-operator namespace
```
Comment (tpantelis) [inline pkg/agent/controller/types.go:32]:
```suggestion
// The ServiceImportController listens for ServiceImport resources created in the target namespace
```

File: pkg/agent/controller/types.go [hunk 30]
```diff
@@ -23,3 +28,36 @@ type AgentSpecification struct {
...
+// to listen only for the endpoints event related to ServiceImport created
+type ServiceImportController struct {
+	kubeClientSet             kubernetes.Interface
+	lighthouseClient          lighthouseClientset.Interface
+	serviceInformer           cache.SharedIndexInformer
+	queue                     workqueue.RateLimitingInterface
+	endpointControllerCreated sync.Map
+	endpointControllerDeleted sync.Map
+	clusterID                 string
+	lighthouseNamespace       string
```
Comment (tpantelis) [inline pkg/agent/controller/types.go:43]:
```suggestion
	namespace       string
```

File: pkg/agent/controller/types.go [hunk 31]
```diff
@@ -23,3 +28,36 @@ type AgentSpecification struct {
...
+}
+
+// Each EndpointController listens for the endpoints that backs a service and have a ServiceImport
+// It will create an endpoint slice corresponding to an endpoint object and set the owner references
+// to ServiceImport. The app label from the endpoint will be added to endpoint slice as well.
+type EndpointController struct {
+	kubeClientSet      kubernetes.Interface
+	endpointInformer   cache.SharedIndexInformer
+	endPointqueue      workqueue.RateLimitingInterface
```
Comment (tpantelis) [inline pkg/agent/controller/types.go:52]:
```suggestion
	endpointQueue      workqueue.RateLimitingInterface
```

File: pkg/agent/controller/endpoint.go [hunk 32]
```diff
@@ -0,0 +1,273 @@
...
+		kubeClientSet:     kubeClientSet,
+		clusterID:         clusterId,
+		stopCh:            make(chan struct{}),
+	}
+
+	return endpointController, nil
+}
+
+func (e *EndpointController) Start(stopCh <-chan struct{}, labelSelector fmt.Stringer) error {
+	e.endpointInformer = cache.NewSharedIndexInformer(
```
Comment (tpantelis) [inline pkg/agent/controller/endpoint.go:39]: You don't need a shared Informer since you have a single dedicated event handler.

```suggestion
	e.endpointInformer = cache.NewInformer(
```

File: pkg/agent/controller/endpoint.go [hunk 33]
```diff
@@ -0,0 +1,273 @@
...
+	endpointSliceName := endPoints.Name + "-" + e.clusterID
+	newEndPointSlice := e.endpointSliceFromEndpoints(endPoints)
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentEndpointSice, err := e.kubeClientSet.DiscoveryV1beta1().EndpointSlices(endPoints.Namespace).
+			Get(endpointSliceName, metav1.GetOptions{})
+
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if currentEndpointSice != nil && currentEndpointSice.ObjectMeta.Name != "" {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpoint.go:157]: Get is retrieving an object with fields having nil values when endpoint slice does not exist, hence added this check.

File: pkg/agent/controller/serviceimport.go (hunk 18 above)
Comment (aswinsuryan) [reply pkg/agent/controller/serviceimport.go:166]: Done.

File: pkg/agent/controller/endpoint.go [hunk 34]
```diff
@@ -0,0 +1,273 @@
...
+			if equality.Semantic.DeepEqual(currentEndpointSice, newEndPointSlice) {
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpoint.go]: Changed to check the interested fields

File: pkg/agent/controller/endpoint.go [hunk 35]
```diff
@@ -0,0 +1,273 @@
...
+			} else {
+				queue.Forget(key)
+			}
+		}()
+	}
+}
+
+func (e *EndpointController) endPointCreatedOrUpdated(obj interface{}, key string) error {
+	endPoints := obj.(*corev1.Endpoints)
+	endpointSliceName := endPoints.Name + "-" + e.clusterID
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpoint.go]: Done.

File: pkg/agent/controller/endpoint.go [hunk 36]
```diff
@@ -0,0 +1,273 @@
...
+
+	endPoints := obj.(corev1.Endpoints)
+	matchLabels := endPoints.ObjectMeta.Labels
+	labelSelector := labels.Set(map[string]string{"app": matchLabels["app"]}).AsSelector()
+	err := e.kubeClientSet.DiscoveryV1beta1().EndpointSlices(endPoints.Namespace).
+		DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector.String()})
+	if err != nil && !errors.IsNotFound(err) {
+		klog.Errorf("Deleting EndpointSlice for endPoints %v from NameSpace %s failed due to %v", endPoints, endPoints.Namespace, err)
+		e.endpointDeletedMap.Store(key, endPoints)
+		e.endPointqueue.AddRateLimited(key)
```
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpoint.go]: Done.

File: pkg/agent/controller/serviceimport.go (hunk 28 above)
Comment (aswinsuryan) [reply pkg/agent/controller/serviceimport.go]: Done.

File: pkg/agent/controller/endpoint.go (hunk 33 above)
[the remark this answers is not in the document]
Comment (aswinsuryan) [reply pkg/agent/controller/endpoint.go]: Oh it was an oversight, I could have used the notFound error to decide whether to create or update. Changed the code.

File: pkg/agent/controller/serviceimport.go [hunk 37]
```diff
@@ -0,0 +1,217 @@
...
+
+			if err != nil {
+				klog.Errorf("Error retrieving the object with store is  %v from the cache: %v", c.serviceInformer.GetIndexer().ListKeys(), err)
+				// requeue the item to work on later
+				c.queue.AddRateLimited(key)
+
+				return
+			}
+
+			c.queue.Forget(key)
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:129]: This is a bit premature as it could get re-queued below. I suggest structuring this like you did in the endpoint controller, ie returning an error from the functions below etc

File: pkg/agent/controller/serviceimport.go [hunk 38]
```diff
@@ -0,0 +1,217 @@
...
+	if err != nil {
+		klog.Errorf("Error creating Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
+		return
+	}
+
+	err = endpointController.Start(endpointController.stopCh, labelSelector)
+	if err != nil {
+		klog.Errorf("Error starting Endpoint controller for service %s/%s: %v", serviceNameSpace, serviceName, err)
+		return
+	}
```
Comment (tpantelis) [inline pkg/agent/controller/serviceimport.go:186]: So if `NewEndpointController` or `Start` return an error should we re-queue? However, they don't actually have code paths that return an error. So I suggest simplifying this and change the signatures to not return an error.

File: pkg/agent/main.go [hunk 39]
```diff
@@ -53,15 +53,23 @@ func main() {
...
+	if err != nil {
+		klog.Fatalf("Failed to create ServiceImport controller: %v", err)
+	}
+
 	if err := lightHouseAgent.Start(stopCh); err != nil {
 		klog.Fatalf("Failed to start lighthouse agent: %v", err)
 	}
 
+	if err := esController.Start(stopCh); err != nil {
+		klog.Fatalf("Failed to start endpointSlice controller: %v", err)
```
Comment (tpantelis) [inline pkg/agent/main.go:70]:
```suggestion
		klog.Fatalf("Failed to start Endpoints controller: %v", err)
```

Comment (aswinsuryan) [reply pkg/agent/main.go]: Done.

File: pkg/agent/controller/serviceimport.go (hunk 38 above)
Comment (aswinsuryan) [reply pkg/agent/controller/serviceimport.go]: Done.

File: pkg/agent/controller/serviceimport.go (hunk 37 above)
Comment (aswinsuryan) [reply pkg/agent/controller/serviceimport.go]: Done.

**Review by vthapar — APPROVED** [review]
LGTM
